lib/core/test/unit/Network/Wai/Middleware/LoggingSpec.hs

```diff
@@ -111,80 +119,114 @@ spec = describe "Logging Middleware"
+            , (Debug, "")
             , (Info, "200 OK")
             , (Debug, "\NUL\NUL\NUL")
+            , (Debug, "LogRequestFinish")
             ]
 
     it "POST, 201, with sensitive fields" $ \ctx -> do
         post ctx "/post" (MkJson { field = "patate", sensitive = 14 })
         expectLogs ctx
-            [ (Info, "[POST] /post")
+            [ (Debug, "LogRequestStart")
+            , (Info, "[POST] /post")
             , (Debug, "{\"sensitive\":\"*****\",\"field\":\"patate\"}")
             , (Info, "201 Created")
             , (Debug, "{\"status\":\"ok\",\"whatever\":42}")
+            , (Debug, "LogRequestFinish")
             ]
 
     it "POST, 400, invalid payload (not json)" $ \ctx -> do
         postIlled ctx "/post" "\NUL\NUL\NUL"
         expectLogs ctx
-            [ (Info, "[POST] /post")
+            [ (Debug, "LogRequestStart")
+            , (Info, "[POST] /post")
             , (Debug, "Invalid payload: not JSON")
             , (Info, "400 Bad Request")
             , (Debug, "Failed reading: not a valid json value")
+            , (Debug, "LogRequestFinish")
             ]
 
     it "DELETE, 202, no query" $ \ctx -> do
         delete ctx "/delete"
         expectLogs ctx
-            [ (Info, "[DELETE] /delete")
+            [ (Debug, "LogRequestStart")
+            , (Info, "[DELETE] /delete")
+            , (Debug, "")
             , (Info, "204 No Content")
+            , (Debug, "")
+            , (Debug, "LogRequestFinish")
             ]
 
     it "GET, 400" $ \ctx -> do
         get ctx "/error400"
         expectLogs ctx
-            [ (Info, "[GET] /error400")
+            [ (Debug, "LogRequestStart")
+            , (Info, "[GET] /error400")
+            , (Debug, "")
             , (Info, "400 Bad Request")
+            , (Debug, "")
+            , (Debug, "LogRequestFinish")
             ]
 
     it "GET, 500" $ \ctx -> do
         get ctx "/error500"
         expectLogs ctx
-            [ (Info, "[GET] /error500")
+            [ (Debug, "LogRequestStart")
+            , (Info, "[GET] /error500")
+            , (Debug, "")
             , (Error, "500 Internal Server Error")
+            , (Debug, "")
+            , (Debug, "LogRequestFinish")
             ]
 
     it "GET, 503" $ \ctx -> do
         get ctx "/error503"
         expectLogs ctx
-            [ (Info, "[GET] /error503")
+            [ (Debug, "LogRequestStart")
+            , (Info, "[GET] /error503")
+            , (Debug, "")
             , (Warning, "503 Service Unavailable")
+            , (Debug, "")
+            , (Debug, "LogRequestFinish")
             ]
 
     it "different request ids" $ \ctx -> property $ \(NumberOfRequests n) ->
         monadicIO $ liftIO $ do
             void $ mapConcurrently (const (get ctx "/get")) (replicate n ())
             entries <- readTVarIO (logs ctx)
-            let index = Map.fromList [ (loName l, l) | l <- entries ]
+            let getReqId l = case loContent l of
```

Comment:
This could still be a list comprehension, right? Failed pattern matches are filtered out.

Reply:
my intention here was to get RequestId from each log, and then nubBy the entries by this requestId. Thanks to that, I am expected to end up with unique logs having distinct RequestId. So...probably can be list comprehension, but I find the proposed version more clear (but this is subjective;-)